File: backend/app/analysis/features.py
"""Acoustic feature extraction: timing, pitch, voice quality, volume, tremor.

Metrics that cannot be measured on a given recording (e.g. jitter on an
unvoiced sample) are returned as None and skipped by the trend analysis.
"""

import numpy as np
import parselmouth
from parselmouth.praat import call
from scipy import signal

from ..config import PAUSE_THRESHOLD_S
from .dsp import frame_rms, mfcc

PITCH_FLOOR = 75
PITCH_CEILING = 500


def extract_features(audio: np.ndarray, sr: int, words: list[dict]) -> dict:
    snd = parselmouth.Sound(audio.astype(np.float64), sampling_frequency=sr)
    features = {
        "duration_s": round(len(audio) / sr, 2),
        **_timing_metrics(words),
        **_pitch_metrics(snd),
        **_voice_quality_metrics(snd),
        "mean_volume_db": _mean_volume_db(audio),
        "tremor_index": _tremor_index(audio, sr),
        "mfcc_means": _mfcc_means(audio, sr),
    }
    return _finite(features)


def _timing_metrics(words: list[dict]) -> dict:
    if len(words) < 2:
        return {
            "word_count": len(words),
            "speech_rate_wpm": 0.0,
            "articulation_rate_wpm": 0.0,
            "pause_count": 0,
            "pauses_per_minute": 0.0,
            "avg_pause_duration_s": 0.0,
            "rhythm_variability": None,
            "pronunciation_confidence": words[0]["probability"] if words else None,
        }
    span = max(words[-1]["end"] - words[0]["start"], 0.1)
    gaps = [nxt["start"] - cur["end"] for cur, nxt in zip(words, words[1:])]
    pauses = [g for g in gaps if g > PAUSE_THRESHOLD_S]
    speaking_time = max(span - sum(pauses), 0.1)
    durations = [w["end"] - w["start"] for w in words]
    mean_duration = float(np.mean(durations))
    return {
        "word_count": len(words),
        "speech_rate_wpm": round(len(words) / span * 60, 1),
        "articulation_rate_wpm": round(len(words) / speaking_time * 60, 1),
        "pause_count": len(pauses),
        "pauses_per_minute": round(len(pauses) / span * 60, 2),
        "avg_pause_duration_s": round(float(np.mean(pauses)), 3) if pauses else 0.0,
        "rhythm_variability": round(float(np.std(durations)) / mean_duration, 3) if mean_duration > 0 else None,
        "pronunciation_confidence": round(float(np.mean([w["probability"] for w in words])), 3),
    }


def _pitch_metrics(snd: parselmouth.Sound) -> dict:
    pitch = snd.to_pitch(time_step=0.01, pitch_floor=PITCH_FLOOR, pitch_ceiling=PITCH_CEILING)
    frequencies = pitch.selected_array["frequency"]
    voiced = frequencies[frequencies > 0]
    if voiced.size < 10:
        return {"mean_pitch_hz": None, "pitch_variability_semitones": None}
    mean_hz = float(np.mean(voiced))
    semitones = 12 * np.log2(voiced / mean_hz)
    return {
        "mean_pitch_hz": round(mean_hz, 1),
        "pitch_variability_semitones": round(float(np.std(semitones)), 2),
    }


def _voice_quality_metrics(snd: parselmouth.Sound) -> dict:
    try:
        point_process = call(snd, "To PointProcess (periodic, cc)", PITCH_FLOOR, PITCH_CEILING)
        jitter = call(point_process, "Get jitter (local)", 0, 0, 0.0001, 0.02, 1.3) * 100
        shimmer = call([snd, point_process], "Get shimmer (local)", 0, 0, 0.0001, 0.02, 1.3, 1.6) * 100
        harmonicity = call(snd, "To Harmonicity (cc)", 0.01, PITCH_FLOOR, 0.1, 1.0)
        hnr = call(harmonicity, "Get mean", 0, 0)
        return {
            "jitter_percent": round(float(jitter), 2),
            "shimmer_percent": round(float(shimmer), 2),
            "hnr_db": round(float(hnr), 1),
        }
    except Exception:
        return {"jitter_percent": None, "shimmer_percent": None, "hnr_db": None}


def _mean_volume_db(audio: np.ndarray) -> float | None:
    """Mean level (dBFS) of the louder frames, so silence doesn't drag it down."""
    rms = frame_rms(audio, frame_length=2048, hop_length=512)
    active = rms[rms > np.max(rms) * 0.1]
    if active.size == 0:
        return None
    return round(float(np.mean(20 * np.log10(active + 1e-10))), 1)


def _tremor_index(audio: np.ndarray, sr: int) -> float | None:
    """Share of amplitude-envelope modulation in the 3-12 Hz band, where
    pathological voice tremor typically appears, relative to 0.5-20 Hz."""
    hop = sr // 100  # 100 Hz envelope
    envelope = frame_rms(audio, frame_length=hop * 4, hop_length=hop)
    if envelope.size < 64:
        return None
    envelope = envelope - np.mean(envelope)
    frequencies, power = signal.welch(envelope, fs=100, nperseg=min(256, envelope.size))
    total = float(power[(frequencies >= 0.5) & (frequencies <= 20)].sum())
    tremor = float(power[(frequencies >= 3) & (frequencies <= 12)].sum())
    if total <= 0:
        return None
    return round(tremor / total, 3)


def _mfcc_means(audio: np.ndarray, sr: int) -> list[float]:
    coefficients = mfcc(audio, sr, n_mfcc=13)
    return [round(float(v), 2) for v in coefficients.mean(axis=0)]


def _finite(features: dict) -> dict:
    """Replace NaN/inf with None so the output is always valid JSON."""
    cleaned = {}
    for key, value in features.items():
        if isinstance(value, float) and not np.isfinite(value):
            cleaned[key] = None
        elif isinstance(value, list):
            cleaned[key] = [v if np.isfinite(v) else None for v in value]
        else:
            cleaned[key] = value
    return cleaned

File: voice_health/lib/analysis/features.dart
/// Acoustic feature extraction — the Dart port of the backend's features.py.
/// Metrics that cannot be measured on a recording are null and are skipped
/// by the trend analysis.
library;

import 'dart:math' as math;
import 'dart:typed_data';

import 'dsp.dart';
import 'pitch.dart';
import 'voice_quality.dart';

const kPauseThresholdS = 0.35;

/// A transcribed word with its position on the recording's timeline.
class WordTiming {
  final String word;
  final double start;
  final double end;

  const WordTiming(this.word, this.start, this.end);
}

Map<String, double?> extractFeatures(
  Float64List x,
  int sr,
  List<WordTiming> words,
) {
  final pitch = trackPitch(x, sr);
  final quality = analyzeVoiceQuality(x, sr, pitch);
  return {
    'duration_s': _round(x.length / sr, 2),
    ..._timingMetrics(words),
    ..._pitchMetrics(pitch),
    'jitter_percent': _round(quality.jitterPercent, 2),
    'shimmer_percent': _round(quality.shimmerPercent, 2),
    'hnr_db': _round(quality.hnrDb, 1),
    'mean_volume_db': _round(_meanVolumeDb(x), 1),
    'tremor_index': _round(_tremorIndex(x, sr), 3),
  };
}

Map<String, double?> _timingMetrics(List<WordTiming> words) {
  if (words.length < 2) {
    return {
      'word_count': words.length.toDouble(),
      'speech_rate_wpm': 0,
      'articulation_rate_wpm': 0,
      'pause_count': 0,
      'pauses_per_minute': 0,
      'avg_pause_duration_s': 0,
      'rhythm_variability': null,
    };
  }
  final span = math.max(words.last.end - words.first.start, 0.1);
  final pauses = <double>[];
  for (var i = 0; i < words.length - 1; i++) {
    final gap = words[i + 1].start - words[i].end;
    if (gap > kPauseThresholdS) pauses.add(gap);
  }
  final pauseTime = pauses.fold(0.0, (a, b) => a + b);
  final speakingTime = math.max(span - pauseTime, 0.1);
  final durations = [for (final w in words) w.end - w.start];
  final meanDuration = durations.reduce((a, b) => a + b) / durations.length;
  var durationVariance = 0.0;
  for (final d in durations) {
    durationVariance += (d - meanDuration) * (d - meanDuration);
  }
  final durationStd = math.sqrt(durationVariance / durations.length);
  return {
    'word_count': words.length.toDouble(),
    'speech_rate_wpm': _round(words.length / span * 60, 1),
    'articulation_rate_wpm': _round(words.length / speakingTime * 60, 1),
    'pause_count': pauses.length.toDouble(),
    'pauses_per_minute': _round(pauses.length / span * 60, 2),
    'avg_pause_duration_s':
        pauses.isEmpty ? 0 : _round(pauseTime / pauses.length, 3),
    'rhythm_variability':
        meanDuration > 0 ? _round(durationStd / meanDuration, 3) : null,
  };
}

Map<String, double?> _pitchMetrics(PitchTrack pitch) {
  final voiced = pitch.voiced;
  if (voiced.length < 10) {
    return {'mean_pitch_hz': null, 'pitch_variability_semitones': null};
  }
  final meanHz = voiced.reduce((a, b) => a + b) / voiced.length;
  var variance = 0.0;
  for (final f in voiced) {
    final semitones = 12 * math.log(f / meanHz) / math.ln2;
    variance += semitones * semitones;
  }
  return {
    'mean_pitch_hz': _round(meanHz, 1),
    'pitch_variability_semitones':
        _round(math.sqrt(variance / voiced.length), 2),
  };
}

/// Mean level (dBFS) of the louder frames, so silence doesn't drag it down.
double? _meanVolumeDb(Float64List x) {
  final rms = frameRms(x, 2048, 512);
  var maxRms = 0.0;
  for (final v in rms) {
    if (v > maxRms) maxRms = v;
  }
  var sum = 0.0, count = 0;
  for (final v in rms) {
    if (v > maxRms * 0.1) {
      sum += 20 * math.log(v + 1e-10) / math.ln10;
      count++;
    }
  }
  return count > 0 ? sum / count : null;
}

/// Share of amplitude-envelope modulation in the 3-12 Hz band, where
/// pathological voice tremor typically appears, relative to 0.5-20 Hz.
double? _tremorIndex(Float64List x, int sr) {
  final hop = sr ~/ 100; // 100 Hz envelope
  final envelope = frameRms(x, hop * 4, hop);
  if (envelope.length < 64) return null;
  var mean = 0.0;
  for (final v in envelope) {
    mean += v;
  }
  mean /= envelope.length;
  final centered = Float64List(envelope.length);
  for (var i = 0; i < envelope.length; i++) {
    centered[i] = envelope[i] - mean;
  }
  final spectrum = welch(centered, 100, nperseg: math.min(256, centered.length));
  var total = 0.0, tremor = 0.0;
  for (var i = 0; i < spectrum.freqs.length; i++) {
    final f = spectrum.freqs[i];
    if (f >= 0.5 && f <= 20) total += spectrum.psd[i];
    if (f >= 3 && f <= 12) tremor += spectrum.psd[i];
  }
  return total > 0 ? tremor / total : null;
}

double? _round(double? value, int decimals) {
  if (value == null || !value.isFinite) return null;
  final factor = math.pow(10, decimals);
  return (value * factor).round() / factor;
}
